samza-sql-avro/src/main/java/org/apache/samza/sql/avro/AvroTypeFactoryImpl.java

```diff
@@ -0,0 +1,133 @@
+        createTypeWithNullability(createSqlType(SqlTypeName.VARCHAR), true)));
+
+    for (Schema.Field field : fields) {
+      String fieldName = field.name();
+      int fieldPos = field.pos() + 1;
+      RelDataType dataType = getRelDataType(field.schema());
+      relFields.add(new RelDataTypeFieldImpl(fieldName, fieldPos, dataType));
+    }
+
+    return relFields;
+  }
+
+  private RelDataType getRelDataType(Schema fieldSchema) {
+    switch (fieldSchema.getType()) {
+      case ARRAY:
+        // JavaTypeFactoryImpl converts Array into Array(ANY, ANY)
+        return new ArraySqlType(createSqlType(SqlTypeName.ANY), true);
+      case BOOLEAN:
+        return createTypeWithNullability(createSqlType(SqlTypeName.BOOLEAN), true);
+      case DOUBLE:
+        return createTypeWithNullability(createSqlType(SqlTypeName.DOUBLE), true);
+      case FLOAT:
+        return createTypeWithNullability(createSqlType(SqlTypeName.FLOAT), true);
+      case ENUM:
+        return createTypeWithNullability(createSqlType(SqlTypeName.VARCHAR), true);
+      case UNION:
+        return getRelTypeFromUnionTypes(fieldSchema.getTypes());
+      case FIXED:
+        return createTypeWithNullability(createSqlType(SqlTypeName.VARBINARY), true);
+      case STRING:
+        return createTypeWithNullability(createSqlType(SqlTypeName.VARCHAR), true);
+      case BYTES:
+        return createTypeWithNullability(createSqlType(SqlTypeName.VARBINARY), true);
+      case INT:
+        return createTypeWithNullability(createSqlType(SqlTypeName.INTEGER), true);
+      case LONG:
+        return createTypeWithNullability(createSqlType(SqlTypeName.BIGINT), true);
+      case RECORD:
+//        List<RelDataTypeField> relFields = getRelFields(fieldSchema);
+//        return new RelRecordType(relFields);
+        // TODO Calcite execution engine doesn't support record type yet.
+        return createTypeWithNullability(createSqlType(SqlTypeName.ANY), true);
+      case MAP:
+        // JavaTypeFactoryImpl converts map into Map(ANY, ANY)
+        return super.createMapType(createTypeWithNullability(createSqlType(SqlTypeName.ANY), true),
+            createTypeWithNullability(createSqlType(SqlTypeName.ANY), true));
+      default:
+        String msg = String.format("Field Type %s is not supported", fieldSchema.getType());
+        LOG.error(msg);
+        throw new SamzaException(msg);
+    }
+  }
+
+  private RelDataType getRelTypeFromUnionTypes(List<Schema> types) {
+    // Typically a nullable field's schema is configured as an union of Null and a Type.
+    // This is to check whether the Union is a Nullable field
+    if (types.size() == 2) {
+      if (types.get(0).getType() == Schema.Type.NULL) {
+        return getRelDataType(types.get(1));
+      } else if ((types.get(1).getType() == Schema.Type.NULL)) {
+        return getRelDataType(types.get(0));
+      }
+    }
+
+    return createTypeWithNullability(createSqlType(SqlTypeName.VARCHAR), true);
```

Comment:
@nickpan47 Schema conversion is just one way (i.e. Avro -> Rel) Only the message conversion happens two ways ( Avro message <=> Rel Message). Since we don't convert the relational schema to avro schema, We really don't convert the nullable varchar to any specific avro type. 
If your question is w.r.t to the dynamic avro schema generation during the rel to avro message conversion? It is just used for testing, it always creates a nullable fields in avro which is basically union of ["null", "string"]